Use 'ShellExecute' instead of 'wsystem' in 'runCommand' for Windows.

darosior opened the pull request:
As pointed out by #15883, the use of `blocknotify` or `walletnotify` on Windows caused a flickering command window at each call.

This PR substitutes the use of `wsystem` in `runCommand` with [createProcess](https://docs.microsoft.com/en-us/windows/desktop/api/processthreadsapi/nf-processthreadsapi-createprocessa), using the [CREATE_NO_WINDOW](https://docs.microsoft.com/en-us/windows/desktop/ProcThread/process-creation-flags) flag to avoid this behavior.

promag commented:
@MarcoFalke can we have a gitian build?

promag commented:
Have you considered using boost process?

Sjors commented:
If you decide to use boost process, consider building on top of #15382 (minus the last commit). That said, this change is much smaller (though I have no idea if it's correct).

MarcoFalke commented:
```
c:\projects\bitcoin\src\util\system.cpp(1140): error C2664: 'BOOL CreateProcessW(LPCWSTR,LPWSTR,LPSECURITY_ATTRIBUTES,LPSECURITY_ATTRIBUTES,BOOL,DWORD,LPVOID,LPCWSTR,LPSTARTUPINFOW,LPPROCESS_INFORMATION)': cannot convert argument 2 from 'char *' to 'LPWSTR

darosior commented:
> ```
> c:\projects\bitcoin\src\util\system.cpp(1140): error C2664: 'BOOL CreateProcessW(LPCWSTR,LPWSTR,LPSECURITY_ATTRIBUTES,LPSECURITY_ATTRIBUTES,BOOL,DWORD,LPVOID,LPCWSTR,LPSTARTUPINFOW,LPPROCESS_INFORMATION)': cannot convert argument 2 from 'char *' to 'LPWSTR
> ```

Fixed it by explicitly using `CreateProcessA` (I used `CreateProcess` that was interpreted as `CreateProcessA` during my tests and as `CreateProcessW` by appveyor).



> Have you considered using boost process?

> If you decide to use boost process, consider building on top of #15382 (minus the last commit). That said, this change is much smaller (though I have no idea if it's correct).

Is it a good idea to add code that use boost since, as I've read [here](https://github.com/bitcoin/bitcoin/projects), we aim to remove Boost as a dependency ?

Sjors commented:
The idea is to move from Boost to modern C++, where it's often the case that new C++ standards adopt things from Boost. So the hope is the boost::process can eventually easily be replaced with something in C++42 :-)

laanwj commented:
This change was rejected before in #5834—not sure the same reasoning, or problems exist with this PR, but have a look.

luke-jr commented on src/util/system.cpp:
```diff
@@ -1128,10 +1128,22 @@ fs::path GetSpecialFolderPath(int nFolder, bool fCreate)
+    }
 #endif
     if (nErr)
         LogPrintf("runCommand error: system(%s) returned %d\n", strCommand, nErr);
```
This message is explicit about the call being `system`, so should probably be made ambiguous

luke-jr commented on src/util/system.cpp:
```diff
@@ -1128,10 +1128,22 @@ fs::path GetSpecialFolderPath(int nFolder, bool fCreate)
+    STARTUPINFOA info;
+    PROCESS_INFORMATION processInfo;
+    ZeroMemory(&info, sizeof(info));
+    info.cb = sizeof(info);
```
Why not just initialise info with `{.cb = sizeof(info)}` above?

luke-jr commented on src/util/system.cpp:
```diff
@@ -1128,10 +1128,22 @@ fs::path GetSpecialFolderPath(int nFolder, bool fCreate)
+    PROCESS_INFORMATION processInfo;
+    ZeroMemory(&info, sizeof(info));
+    info.cb = sizeof(info);
+    ZeroMemory(&processInfo, sizeof(processInfo));
```
Is there a need to zero this?

darosior replied:
I had never used the windows API before this, so I sticked to the documentation example : https://docs.microsoft.com/en-us/windows/desktop/ProcThread/creating-processes

luke-jr commented on src/util/system.cpp:
```diff
@@ -1128,10 +1128,22 @@ fs::path GetSpecialFolderPath(int nFolder, bool fCreate)
+    ZeroMemory(&info, sizeof(info));
+    info.cb = sizeof(info);
+    ZeroMemory(&processInfo, sizeof(processInfo));
+    if (CreateProcessA(NULL, const_cast<char *>(strCommand.c_str()), NULL, NULL, FALSE, 0x08000000, NULL, NULL, &info, &processInfo)) {
```
`CreateProcessA` seems to execute the program directly, rather than passing through the shell as expected for `-*notify`. This may break existing usage, especially because `.bat` files will fail.

darosior replied:
According to the [doc](https://docs.microsoft.com/en-us/windows/desktop/api/processthreadsapi/nf-processthreadsapi-createprocessa), "The CreateProcess function creates a new process, which runs independently of the creating process.".

darosior replied:
Indeed, a workaround is to put `cmd.exe` at the command beginning as in #5834, but I don't know if this is a good way to do it..

luke-jr commented on src/util/system.cpp:
```diff
@@ -1128,10 +1128,22 @@ fs::path GetSpecialFolderPath(int nFolder, bool fCreate)
+    if (CreateProcessA(NULL, const_cast<char *>(strCommand.c_str()), NULL, NULL, FALSE, 0x08000000, NULL, NULL, &info, &processInfo)) {
+        WaitForSingleObject(processInfo.hProcess, INFINITE);
+        CloseHandle(processInfo.hProcess);
+        CloseHandle(processInfo.hThread);
```
Where is `nErr` set to the process's return value?

darosior replied:
`nErr` is set only if the command did not succeed. Maybe it should be set in any case ?

darosior replied:
nErr is set only if the command did not succeed.

luke-jr replied:
`nErr` should always be set to the exit code of the process.

darosior replied:
I could have set `nErr` to `GetLastError()` outside of the `else` block but I've read from [the doc](https://msdn.microsoft.com/fr-fr/library/windows/desktop/ms679360(v=vs.85).aspx) that a windows process would not always return `0` on success : "some functions set the last-error code to 0 on success and others do not." This means that `if (nErr)` could pass and log as an error on windows even if the called process executed successfully.
I can change it (along with the log message as asked above) but I'll first give a look at using `ShellExecute` as Sipa proposed in #5834.

luke-jr reviewed (changes requested):
I'm concerned that this may cause output from executed commands to be lost when the user is running a non-daemon bitcoind. Can anyone test?

luke-jr commented:
re boost::process, note that early versions had bugs which IIRC would likely break this use case.

darosior commented:
@luke-jr thank you for the review. I finally opted for the `ShellExecute` function to handle batch files, as proposed by Sipa in #5834 (thank you @laanwj for pointing me to this PR), and to reduce diff. I also changed the Log message and not set `nErr` to the process exit code because [ShellExecute can return 0 on error](https://docs.microsoft.com/en-us/windows/desktop/api/shellapi/nf-shellapi-shellexecutea#return-value).

luke-jr commented on src/util/system.cpp:
```diff
@@ -1130,11 +1130,13 @@ void runCommand(const std::string& strCommand)
         LogPrintf("runCommand error: system(%s) returned %d\n", strCommand, nErr);
+#else
+    // Second argument "verb" (=action to be performed) set to NULL results in action "open"
+    if (!ShellExecute(NULL, NULL, strCommand.c_str(), NULL, NULL, SW_HIDE))
```
Does this actually work at all? From the docs I would expect it doesn't.

darosior replied:
Yes it does, I tested it before comitting.

darosior replied:
Actually you are right about the condition I messed up the return values of [ShellExecuteEx](https://docs.microsoft.com/en-us/windows/desktop/api/Shellapi/nf-shellapi-shellexecuteexa) (boolean) and [ShellExecute](https://docs.microsoft.com/en-us/windows/desktop/api/shellapi/nf-shellapi-shellexecutea#return-value) (HINSTANCE). I am going to fix it right now.

NicolasDorier commented:
You might need to call `CreateProcessW` or `CreateProcess` instead ping @ken2812221 . I am unsure.

darosior committed:
Use 'ShellExecuteEx' instead of 'wsystem' in 'runCommand' for Windows

'blocknotify' or 'walletnotify' on Windows caused a cmd window to popup. The use of 'ShellExecute' with the 'SW_HIDE' flag instead of 'wsystem' avoid this behaviour

fanquake commented:
@ken2812221 Is not currently available. So maybe we can defer to @sipsorcery for a Concept ACK / NACK.

sipsorcery commented:
@darosior if the command can ever contain unicode characters then `SHELLEXECUTEINFOW` should be used instead of `SHELLEXECUTEINFOA`. 

@ken2812221's work added some big improvements to the bitcoin core software unicode handling on Windows. The `datadir` and other paths should be treated as potentially containing unicode characters.

darosior commented:
Thank you for the help with the Windows version which I struggled to test. I'll revisit this PR soon.

luke-jr commented on src/util/system.cpp:
```diff
@@ -1130,11 +1130,22 @@ void runCommand(const std::string& strCommand)
+    ShExecInfo.lpParameters = NULL;
+    ShExecInfo.lpDirectory = NULL;
+    ShExecInfo.nShow = SW_HIDE;
+    ShExecInfo.hInstApp = NULL;
```
```c++
    SHELLEXECUTEINFOA ShExecInfo = {
        .cbSize = sizeof(SHELLEXECUTEINFOA),
        .lpFile = strCommand.c_str(),
        .nShow = SW_HIDE,
    };
```

luke-jr commented on src/util/system.cpp:
```diff
@@ -1130,11 +1130,22 @@ void runCommand(const std::string& strCommand)
+    ShExecInfo.lpFile = strCommand.c_str();
+    ShExecInfo.lpParameters = NULL;
+    ShExecInfo.lpDirectory = NULL;
+    ShExecInfo.nShow = SW_HIDE;
```
It shouldn't be hidden if there is already a console available (eg, if bitcoind is being used in a command shell)

luke-jr commented on src/util/system.cpp:
```diff
@@ -1130,11 +1130,22 @@ void runCommand(const std::string& strCommand)
+    ShExecInfo.lpDirectory = NULL;
+    ShExecInfo.nShow = SW_HIDE;
+    ShExecInfo.hInstApp = NULL;
+    if (!ShellExecuteExA(&ShExecInfo))
```
Needs braces

darosior commented:
@luke-jr thank you for the time you spent reviewing this again. However I can't test this PR since I can't access a Windows machine and the appveyor testing is really not viable.

I'm sorry for people experiencing #15883 but I shouldn't have done a Windows PR in the first place (maybe someone can grab and correct this if it's not too wrong) ... Closing.
